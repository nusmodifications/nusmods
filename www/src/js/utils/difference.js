// @flow
import { transform, isEqual, isObject, omit, merge, mapValues } from 'lodash';

const DIFF_PATH = 'DIFF_PATH';
const DIFF_OBJ = 'DIFF_OBJ';

export type PathDiff = {
  type: 'DIFF_PATH',
  previous: Object,
  next: Object,
};

export type ObjectDiff = {
  type: 'DIFF_OBJ',
  deleted: Object,
  added: Object,
  changed: { [any]: PathDiff | ObjectDiff },
};

/*
 * Deep diff between two object, using lodash
 * Based on https://gist.github.com/Yimiprod/7ee176597fef230d1451
 * @param  {Object} object Object compared
 * @param  {Object} base   Object to compare with
 * @return {Object}        Return a new object who represent the diff
 */
function changes(object: Object, base: Object, diffFn: Function): Object {
  return transform(object, (result: Object, value: Object, key: string) => {
    if (!isEqual(value, base[key])) {
      // eslint-disable-next-line no-param-reassign
      result[key] =
        isObject(value) &&
        isObject(base[key]) &&
        !(value instanceof Array) &&
        !(base[key] instanceof Array)
          ? diffFn(value, base[key])
          : { type: DIFF_PATH, previous: base[key], next: value };
    }
  });
}

export function difference(object: Object, base: Object): ObjectDiff {
  return {
    type: DIFF_OBJ,
    deleted: omit(base, Object.keys(object)),
    added: omit(object, Object.keys(base)),
    changed: changes(object, base, difference),
  };
}

// Revert object back to base using diff
export function undoDifference(object: Object, diff: ObjectDiff) {
  let base = object;

  // Remove added keys
  base = omit(base, Object.keys(diff.added));

  // Add deleted keys back
  base = merge(base, diff.deleted);

  // Apply PathDiffs here and ObjectDiffs recursively
  base = mapValues(base, (value: PathDiff | ObjectDiff, key: string) => {
    if (!(key in diff.changed)) return value;
    const valueDiff = diff.changed[key];
    switch (valueDiff.type) {
      case DIFF_PATH:
        return valueDiff.previous;
      case DIFF_OBJ:
        return undoDifference(value, valueDiff);
      default:
        return value;
    }
  });

  return base;
}

// Revert object back from base using diff
export function redoDifference(base: Object, diff: ObjectDiff) {
  let object = base;

  // Remove deleted keys
  object = omit(object, Object.keys(diff.deleted));

  // Add added keys back
  object = merge(object, diff.added);

  // Apply PathDiffs here and ObjectDiffs recursively
  object = mapValues(object, (value: PathDiff | ObjectDiff, key: string) => {
    if (!(key in diff.changed)) return value;
    const valueDiff = diff.changed[key];
    switch (valueDiff.type) {
      case DIFF_PATH:
        return valueDiff.next;
      case DIFF_OBJ:
        return redoDifference(value, valueDiff);
      default:
        return value;
    }
  });

  return object;
}
