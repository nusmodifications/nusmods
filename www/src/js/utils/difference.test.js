// @flow
import { difference, undoDifference, redoDifference } from './difference';

const base = {
  changed: 'ARRAYSSTARTAT1',
  arr: [1, 2],
  untouched: 'untouched',
  deleted: 'deleted',
  deep: {
    obj: {
      one: 'two', // Changed
      unchanged: 'unchanged',
      deleted: 'deleted',
    },
  },
  unchangedObj: {
    twinkle: 'twinkle',
  },
};

const object = {
  changed: 'ARRAYSSTARTAT0', // Changed
  arr: [0, 1], // One elem removed, one added
  untouched: 'untouched',
  added: 'added',
  deep: {
    obj: {
      one: 'buckle', // Changed
      unchanged: 'unchanged',
      added: 'added',
    },
  },
  unchangedObj: {
    twinkle: 'twinkle',
  },
};

describe('#difference()', () => {
  test('should compute all deleted, added and changed object key paths', () => {
    const diff = difference(object, base);
    expect(diff).toMatchSnapshot();
  });
});

describe('#undoDifference()', () => {
  test('should undo all differences', () => {
    const diff = difference(object, base);
    const generatedBase = undoDifference(object, diff);
    expect(generatedBase).toEqual(base);
  });
});

describe('#redoDifference()', () => {
  test('should redo all differences', () => {
    const diff = difference(object, base);
    const generatedObject = redoDifference(base, diff);
    expect(generatedObject).toEqual(object);
  });
});
