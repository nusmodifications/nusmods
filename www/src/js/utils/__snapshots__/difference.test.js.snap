// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#difference() should compute all deleted, added and changed object key paths 1`] = `
Object {
  "added": Object {
    "added": "added",
  },
  "changed": Object {
    "added": Object {
      "next": "added",
      "previous": undefined,
      "type": "DIFF_PATH",
    },
    "arr": Object {
      "next": Array [
        0,
        1,
      ],
      "previous": Array [
        1,
        2,
      ],
      "type": "DIFF_PATH",
    },
    "changed": Object {
      "next": "ARRAYSSTARTAT0",
      "previous": "ARRAYSSTARTAT1",
      "type": "DIFF_PATH",
    },
    "deep": Object {
      "added": Object {},
      "changed": Object {
        "obj": Object {
          "added": Object {
            "added": "added",
          },
          "changed": Object {
            "added": Object {
              "next": "added",
              "previous": undefined,
              "type": "DIFF_PATH",
            },
            "one": Object {
              "next": "buckle",
              "previous": "two",
              "type": "DIFF_PATH",
            },
          },
          "deleted": Object {
            "deleted": "deleted",
          },
          "type": "DIFF_OBJ",
        },
      },
      "deleted": Object {},
      "type": "DIFF_OBJ",
    },
  },
  "deleted": Object {
    "deleted": "deleted",
  },
  "type": "DIFF_OBJ",
}
`;
